contract/models/contract_line.py

```diff
@@ -0,0 +1,819 @@
+        self.ensure_one()
+        invoice_line = self.env['account.invoice.line'].new(
+            {
+                'invoice_id': invoice_id,
+                'product_id': self.product_id.id,
+                'quantity': self.quantity,
+                'uom_id': self.uom_id.id,
+                'discount': self.discount,
+            }
+        )
+        # Get other invoice line values from product onchange
+        invoice_line._onchange_product_id()
+        invoice_line_vals = invoice_line._convert_to_write(invoice_line._cache)
+        # Insert markers
+        contract = self.contract_id
+        lang_obj = self.env['res.lang']
+        lang = lang_obj.search([('code', '=', contract.partner_id.lang)])
+        date_format = lang.date_format or '%m/%d/%Y'
+        name = self._insert_markers(date_format)
+        invoice_line_vals.update(
+            {
+                'name': name,
+                'account_analytic_id': contract.id,
+                'price_unit': self.price_unit,
+            }
+        )
+        return invoice_line_vals
+
+    @api.multi
+    def _insert_markers(self, date_format):
+        self.ensure_one()
+        date_from = fields.Date.from_string(self.recurring_next_date)
+        date_to = date_from + self.get_relative_delta(
+            self.recurring_rule_type, self.recurring_interval
+        )
+        name = self.name
+        name = name.replace('#START#', date_from.strftime(date_format))
+        name = name.replace('#END#', date_to.strftime(date_format))
+        return name
+
+    @api.multi
+    def _update_recurring_next_date(self):
+        for line in self:
+            ref_date = line.recurring_next_date or fields.Date.today()
+            old_date = fields.Date.from_string(ref_date)
+            new_date = old_date + self.get_relative_delta(
+                line.recurring_rule_type, line.recurring_interval
+            )
+            line.recurring_next_date = new_date
+
+    @api.model
+    def get_relative_delta(self, recurring_rule_type, interval):
+        if recurring_rule_type == 'daily':
+            return relativedelta(days=interval)
+        elif recurring_rule_type == 'weekly':
+            return relativedelta(weeks=interval)
+        elif recurring_rule_type == 'monthly':
+            return relativedelta(months=interval)
+        elif recurring_rule_type == 'monthlylastday':
+            return relativedelta(months=interval, day=31)
+        else:
+            return relativedelta(years=interval)
+
+    @api.multi
+    def delay(self, delay_delta):
+        """
+        Delay a contract line
+        :param delay_delta: delay relative delta
+        :return: delayed contract line
+        """
+        for rec in self:
+            old_date_start = rec.date_start
+            old_date_end = rec.date_end
+            new_date_start = rec.date_start + delay_delta
+            rec.recurring_next_date = self._compute_first_recurring_next_date(
```

Comment:
Since this calls _compute_first_recurring_next_date, is it correct that we are not supposed to call delay if the first invoice has been sent?

Reply:
yes,